model/src/CMakeLists.txt

```diff
@@ -0,0 +1,224 @@
+
+
+# Parse switch file and add compile definitions
+set(switch_file ${CMAKE_SOURCE_DIR}/model/bin/switch_${SWITCH})
+if(NOT EXISTS ${switch_file})
+  message(FATAL_ERROR "Switch file '${switch_file}' does not exist, set switch with -DSWITCH=<switch>")
+endif()
+
+file(STRINGS ${switch_file} switch_strings)
+separate_arguments(switches UNIX_COMMAND ${switch_strings})
+
+function(len_minus_one json_array len)
+  
+endfunction()
+
+#TODO check for conflicts and required associated switches
+# Read JSON file
+file(READ ${CMAKE_SOURCE_DIR}/model/bin/switches.json json_str)
+# Get length of top-level array of all switch categories
+string(JSON len LENGTH ${json_str})
+# CMake's foreach RANGE is inclusive, so subtract 1 when looping
+math(EXPR len "${len} - 1")
+
+# Loop over switch categories
+set(switch_files "")
```

Comment:
You have a clever solution to a complex problem, but I suspect the problem is more complex than it needs to be. It's unlikely that all these compiler switches are really needed. Most packages get by without anything like this level of complexity of build switches.

I would suggest that the coding team not get too comfortable with the JSON solution, and instead seek to reduce the number of switches to a manageable level (less than 5 or 6), and then handle them in the usual way in CMake. This will take some time, no doubt.

In the meantime, there should be very strict review of attempts to add more switches.

Reply:
Fair enough @edwardhartnett but we need to be able to compile the code as it is now, in all its complexity.  The solution @kgerheiser came up with is greatly appreciated.

Reply:
I totally agree that @kgerheiser cannot resolve these issues, and the build has to cope with them right now. His solution is elegant.

It's an unfortunate side effect of Kyle's clever solution that it's now easier than ever for the development team to add new switches. However, as we have seen, switches cause build system complexity and test difficulties. Better that if-statements are used in the code then #ifdef in the pre-processor, except for certainly well-understood cases, like whether or not a library is present at build-time.

@JessicaMeixner-NOAA perhaps you should consider some overall configuration namelist that can be read at application startup and can carry the switch settings for those switches that can be handled by code, instead of by the build.

Ideally, at some future point, Kyle's JSON solution could be removed, and the remaining switches handled by CMake command line switches, in the typical manner.

Reply:
Yes the switch file has become a proxy to create multiple combinations of possibilities within the wave model without a review of it being needed or not. The wave model will have to carry out a longer term refactoring/restrategizing on reducing the number of switch options. That will probably take a while :)